### PR TITLE
Disable global request cache when QueryRenderer runs on a server

### PR DESCRIPTION
**Pull request by psaia**
This is an immediate fix to the issue discussed here: https://github.com/facebook/relay/issues/2566

### CODE REVIEW

**Comment by snrbrnjna** [comment]
I can confirm, that this PR fixes the problems with relay when QueryRenderer runs on the server. But the fix is incomplete: in some environments or circumstances, there is a `window` object in the server. A 100% reliable version to check if the runtime is nodejs or a browser is this:

`const isBrowser=new Function("try {return this===window;}catch(e){ return false;}");`
For details on this, see https://stackoverflow.com/a/31090240

So please @psaia modify your PR to check if it is run not in the browser with that approach. Apart from that, I hope this will make it soon into an offical release. I assume until then SSR with relay isn't working out for anyone out there.

**Comment by snrbrnjna** [comment]
@facebook-github-bot why are the Checks for this PR failing?

**Comment by babangsund** [comment]
You can look at the test results. For example: https://travis-ci.org/facebook/relay/jobs/597863063#L2389

**Comment by psaia** [comment]
@snrbrnjna Thanks for that. I've pushed an update.

**Comment by psaia** [comment]
I finally had some time to dig into these failing tests. They're passing now. Some notes:

- I used the `new Function` trick you referenced @snrbrnjna 
- The request cache will be disabled when running on a server but **not** when jest runs so the tests simulate the browser as much as possible. This is what caused the tests to fail.

**Comment by psaia** [comment]
@babangsund Any chance you could help get this merged? It would greatly speed up our builds (since we wouldn't need to use my fork).

**Comment by marc-se** [comment]
Will this be merged on Christmas? 🎄🎅🏼
I wish for nothing more.

**Comment by mattkrick** [comment]
@psaia do you call a fetch function on the server?
In my use case there were 2 retainers holding onto the queryFetcher. The 2nd fix was to replace `Map` with `WeakMap` here:
https://github.com/facebook/relay/blob/master/packages/relay-runtime/query/fetchQueryInternal.js#L37

curious how you guys are doing it!

**Comment by psaia** [comment]
@mattkrick Not directly, no. On the server `<QueryRenderer />` from `react-relay` is abstracting the calls (just as it does on the client) - which obviously wraps `fetchQuery`. There isn't much server-specific code. 

However, your fix looks interesting. I'm curious why a "2nd fix" was necessary? What was the side effect you were seeing that the downstream Map/cache was causing? After debugging, it looked as though the issues discussed in #2566 were specifically caused by this `ReactRelayQueryRenderer` component. Though, if there's an SSR issue with `fetchQuery` all higher level components would be effected as well.

Unrelated, but I'm curious if Facebook plans on deprecated this `ReactRelayQueryRenderer` entirely in favor of hook components which is why there's a lack of urgency here.

We have a single wrapper which abstracts the query renderer at the moment since I figure this API could change at any moment:

```typescript
export default function withData<T>(
  ComposedComponent,
  query: T,
  variables: any = {}
) {
  const relayContext = useContext(ReactRelayContext);
  return (
    <QueryRenderer<T>
      fetchPolicy={"store-and-network"}
      environment={relayContext.environment}
      query={query}
      variabley={variables}
      render={({ error, props }) => {
        if (error) {
          return <div>{error.message}</div>;
        } else if (props) {
          return (
            <ComposedComponent
              {...props}
              environment={relayContext.environment}
            />
          );
        } else {
          return <div>Loading</div>;
        }
      }}
    />
  );
}
```

**Comment by mattkrick** [comment]
@psaia the only answer I have for "why the 2nd fix" is the heap dump told me so 🙈 
Screen Shot 2019-12-13 at 2 58 46 AM

Maybe we SSR differently? Mine is a 3-step process:
```js
const environment = new ServerEnvironment()
const element = <Foo environment={environment}/>
ReactDOMServer.renderToStaticMarkup(element) // 1) attempt to render, relay inits fetches
await environment.load() // 2) wait for fetches to turn into results
return ReactDOMServer.renderToStaticMarkup(element) // 3) render again, now with results
```

My environment for SSR is basically this:
```js
class ServerEnvironment extends Environment {
  async load() {
    if (!this.isFetched) {
      this.isFetched = true
      this.result = await this.request
    }
  }

  fetch = (request, variables) => {
    if (!this.isFetched) {
      this.request = callGraphQLServer(request, variables)
    } else {
      return this.result
    }
  }
}
```

**Comment by psaia** [comment]
@mattkrick Mine is similar except for using [react-relay-network-modern](https://github.com/relay-tools/react-relay-network-modern-ssr/blob/master/src/server.js). Here's a [consolidated gist](https://gist.github.com/psaia/4c3800e58a66a3e787442d58d03ed436).

**Comment by mattkrick** [comment]
@psaia looks pretty similar! not readily apparent how you avoided that internal fetch function, but glad it didn't hit ya! if you'd like to add the `WeakMap` addition to your PR i wouldn't be opposed, otherwise i'm happy to submit one later

**Comment by psaia** [comment]
@mattkrick I'd like to fully understand the double fetch issue. Did this happen on the client or server or both? 

Currently there is a double fetch by design (at least it works well for me): one on the server, another post-hydration on the client.

**Comment by mattkrick** [comment]
@psaia 
in my case i'm not rehydrating, i'm just using it to create some HTML for an email.
You can reproduce this on the client, too though. Provided you create a new environment when someone logs out & logs back in, each new environment will get stored in `requestCachesByEnvironment`. Today, you can mitigate that problem by just clearing the store on logout & reusing the environment container.

here's the internal flow (as i understand it)
you pass in a fetch function to the network layer.
when relay determines it needs to fetch, it checks to see if that fetch is a dupe (e.g. it's already in flight). to do this, it looks up all requests for that particular environment in a map called `requestCachesByEnvironment`.

IMO that requestCache should probably be kept on the environment itself, but that PR would be more than 1 line 😄

**Comment by psaia** [comment]
@mattkrick Thanks for explaining. That makes sense. I could see that causing a memory leak as well.

I was able to reproduce and test so I went ahead and committed.

**Comment by psaia** [comment]
@mattkrick Thanks for explaining. That makes sense. I could see that causing a serious memory leak.

I was able to reproduce and test so I went ahead and committed.

**Comment by ccemeraldeyes** [comment]
When will this be merged? It's passing tests and a desired feature.

File: packages/relay-runtime/query/fetchQueryInternal.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ type RequestCacheEntry = {|
   +subscription: Subscription,
 |};
 
-const requestCachesByEnvironment = new Map();
+const requestCachesByEnvironment = new WeakMap();
```
Comment (maraisr) [inline packages/relay-runtime/query/fetchQueryInternal.js]: This should be wrapped in a `WEAKMAP_SUPPORTED` if block. 

cc  @ccemeraldeyes @psaia

**Comment by maraisr** [comment]
I don't think this should be merged - the reuse of an environment, across SSR requests, is a bad idea, philosophically. I believe the _memory leak_ issue with the retainers come about from using `Maps` to that hold onto Environments. If we can garbage collect those environments, the store naturally get's evicted. Rather than now wrapping a `isBrowser` block.

Please see my PR @ #3014 which has solved this SSR problem for us, as we've been running a fork with that implemented.

cc @ccemeraldeyes @psaia

**Pull request closed without merging**